rs/registry/helpers/src/firewall.rs

```diff
@@ -85,6 +94,56 @@ impl<T: RegistryClient + ?Sized> FirewallRegistry for T {
             .collect();
         Ok(Some(result))
     }
+
+    /// Get the IP addresses of all system subnet nodes in the registry, for endpoints used for core protocol services (p2p, xnet, api)
+    fn get_system_subnet_nodes_ip_addresses(
+        &self,
+        version: RegistryVersion,
+    ) -> RegistryClientResult<Vec<IpAddr>> {
+        let system_subnet_node_ids: Vec<NodeId> = self
+            .get_system_subnet_ids(version)?
+            .map(|system_subnet_ids| {
+                system_subnet_ids
+                    .into_iter()
+                    .filter_map(
+                        |subnet_id| match self.get_node_ids_on_subnet(subnet_id, version) {
+                            Ok(Some(node_ids)) => Some(node_ids),
+                            Ok(None) => None,
+                            Err(_) => None,
+                        },
+                    )
+                    .flatten()
+                    .collect()
+            })
+            .unwrap_or_default();
+
+        let system_subnet_node_ips: Vec<IpAddr> = system_subnet_node_ids
+            .into_iter()
+            .filter_map(|node_id| {
+                match deserialize_registry_value::<NodeRecord>(
+                    self.get_value(&make_node_record_key(node_id), version),
+                ) {
+                    Ok(Some(node_record)) => {
+                        let mut endpoints: Vec<ConnectionEndpoint> = Vec::new();
+                        if let Some(xnet_record) = node_record.xnet {
+                            endpoints.push(xnet_record)
+                        };
+                        if let Some(http_record) = node_record.http {
+                            endpoints.push(http_record)
+                        };
+                        Some(endpoints)
```

Comment:
Same as above, you could just return endpoints, then flattening won't be necessary.